_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,443 @@
+
+```properties
+# ...
+replica.lag.time.max.ms
+# ...
+```
+
+This is the time to replicate a message to all in-sync replicas and return an acknowledgment to a producer. If a follower fails to make a fetch request and catch up with the latest message within the specified lag time, it is removed from in-sync replicas. The time you use depends on both network latency and broker disk bandwidth. You can reduce the lag time to detect failed replicas sooner, but by doing so you might increase the number of followers that fall out of sync needlessly.
+
+>Followers operate only to replicate messages from the partition leader and allow recovery should the leader fail. Followers do not normally serve clients, though [rack configuration](https://strimzi.io/docs/operators/latest/using.html#type-Rack-reference) allows a consumer to consume messages from the closest replica when a Kafka cluster spans multiple datacenters.
+
+A failed leader affects the balance of a Kafka cluster, as does the assignment of partition replicas to brokers. Rebalancing ensures that leaders are evenly distributed across brokers and brokers are not overloaded. You can control the frequency, in seconds, of rebalance checks and the maximum percentage of imbalance allowed for a broker before a rebalance is triggered.
+
+```properties
+#...
+auto.leader.rebalance.enable=true
+leader.imbalance.check.interval.seconds=300
+leader.imbalance.per.broker.percentage=10
+#...
+```
+
+The percentage imbalance for a broker is the gap between the current number of partition leaders it holds and the number of partitions which are preferred leaders. The first broker in a partition’s list of replicas is known as the preferred leader. You can set the percentage to zero to ensure that preferred leaders are always elected.
+
+If the checks for rebalances need more control, you can disable automated rebalances. You can then choose when to trigger a rebalance using the `kafka-leader-election.sh` command line tool.
+Alternatively, you can [use Cruise Control for Strimzi](https://strimzi.io/docs/operators/latest/using.html#cruise-control-concepts-str) to change partition leadership and rebalance replicas across your Kafka cluster in a more intelligent way.
+
+>The Grafana dashboards provided with Strimzi show metrics for under-replicated partitions and partitions that do not have an active leader.
+
+### Unclean leader election
+
+Leader election to an in-sync replica is considered _clean_ because it guarantees no loss of data. And this is what happens by default. Kafka waits until the original leader is back online before messages are picked up again.
+
+But what if there is no in-sync replica to take on leadership? If a minimum number of in-sync replicas is not set, and there are no followers in sync with the partition leader when its hard drive fails irrevocably, data is already lost. Not only that, but a new leader cannot be elected because there are no in-sync followers.
```

Comment:
```suggestion
But what if there is no in-sync replica to take on leadership, for example because the ISR only contained the leader when the leader's disk died? If a minimum number of in-sync replicas is not set, and there are no followers in sync with the partition leader when its hard drive fails irrevocably, data is already lost. Not only that, but _a new leader cannot be elected_ because there are no in-sync followers.
```